src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -85,6 +85,12 @@ std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
 void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
   using ::grpc::health::v1::HealthCheckResponse;
 
+  // If current context is requested to stop, directly destruct itself and exit.
+  if (stopped_) {
+    delete this;
+    return;
```

Comment:
Actually is it OK to delete this while we are inside method of this?

Reply:
Yes, a few examples,
- Just a few lines below: https://github.com/ray-project/ray/blob/2e4a126deecf84d2930ad772e763e206e2f5f7d2/src/ray/gcs/gcs_server/gcs_health_check_manager.cc#L103-L106
- Grpc official async example: https://github.com/grpc/grpc/blob/bac8e34abb13f74fb36c4d4ca0288950aecc4380/examples/cpp/helloworld/greeter_async_server.cc#L119-L122

Theoretically, it's possible to avoid raw pointer with shared pointer, but you need to transfer the ownership to callback.
As for your question, `delete this` is completely legal under harsh condition, that is: no further statement after deletion.